Convex hull in one dimension

```julia
julia> convex_hull([[-1.0], [0.0], [-4.0], [5.0]])
ERROR: AssertionError: this function can only be used for 2-dimensional point clouds; try converting the points to a `VPolytope` first
Stacktrace:
 [1] #convex_hull!#109(::String, ::Function, ::Array{Array{Float64,1},1}) at /Users/forets/.julia/dev/LazySets/src/concrete_convex_hull.jl:77
 [2] #convex_hull#108 at ./none:0 [inlined]
 [3] convex_hull(::Array{Array{Float64,1},1}) at /Users/forets/.julia/dev/LazySets/src/concrete_convex_hull.jl:45
 [4] top-level scope at none:0
```

--- a/docs/src/lib/operations.md
+++ b/docs/src/lib/operations.md
@@ -100,7 +100,6 @@ Inherited from [`LazySet`](@ref):
 
 ```@docs
 convex_hull
-convex_hull!
 right_turn
 monotone_chain!
 ```

--- a/src/concrete_convex_hull.jl
+++ b/src/concrete_convex_hull.jl
@@ -102,7 +102,16 @@ function convex_hull!(points::Vector{VN};
 
     # two points
     if m == 2
-        if n == 1 || n == 2
+        if n == 1
+            p1, p2 = points[1], points[2]
+            if p1 == p2  # check for redundancy
+                pop!(points)
+            elseif p1[1] <= p2[1]
+                nothing
+            else
+                points[1], points[2] = p2, p1
+            end
+        elseif n == 2
             # special case, see #876
             p1, p2 = points[1], points[2]
             if p1 == p2  # check for redundancy
@@ -144,7 +153,8 @@ function _convex_hull_nd!(points::Vector{VN};
         backend = default_polyhedra_backend(V, N)
     end
     Vch = remove_redundant_vertices(V, backend=backend)
-    return vertices_list(Vch)
+    # NOTE: the order of the points in Vch is lost, in general, by using filter!
+    return filter!(pi -> pi ∈ vertices_list(Vch), points)
 end
 
 function _convex_hull_2d!(points::Vector{VN};
@@ -157,17 +167,6 @@ function _convex_hull_2d!(points::Vector{VN};
     else
         error("the convex hull algorithm $algorithm is unknown")
     end
-end
-
-function convex_hull(P1::Vector{N},
-                     P2::HPoly{N};
-                     backend=default_polyhedra_backend(P1, N)) where {N}
-    @assert isdefined(@__MODULE__, :Polyhedra) "this function "*
-        "needs the package 'Polyhedra'"
-    Pch = convexhull(polyhedron(P1; backend=backend),
-                     polyhedron(P2; backend=backend))
-    removehredundancy!(Pch)
-    return convert(typeof(P1), Pch)
 end
 
 """

--- a/src/concrete_convex_hull.jl
+++ b/src/concrete_convex_hull.jl
@@ -1,21 +1,57 @@
-"""
-    convex_hull(points::Vector{VN}; [algorithm]::String="monotone_chain"
-               )::Vector{VN} where {N<:Real, VN<:AbstractVector{N}}
+function default_convex_hull_algorithm(points)
+    if length(points) != 0 && length(first(points)) == 2
+        return "monotone_chain"
+    else
+        return nothing
+    end
+end
 
-Compute the convex hull of points in the plane.
+"""
+    convex_hull(points::Vector{VN};
+                [algorithm]=default_convex_hull_algorithm(points),
+                [backend]=nothing
+                )::Vector{VN} where {N<:Real, VN<:AbstractVector{N}}
+
+Compute the convex hull of the given points.
 
 ### Input
 
-- `points`    -- list of 2D vectors
+- `points`    -- list of vectors
 - `algorithm` -- (optional, default: `"monotone_chain"`) the convex hull
-                 algorithm, valid options are:
-
-    * `"monotone_chain"`
-    * `"monotone_chain_sorted"`
+                 algorithm, see valid options below
+- `backend`   -- (optional, default: `"nothing"`) polyhedral computation backend
+                 for higher-dimensional point sets
 
 ### Output
 
-The convex hull as a list of 2D vectors with the coordinates of the points.
+The convex hull as a list of vectors with the coordinates of the points.
+
+### Algorithm
+
+A pre-processing step treats the cases with `0`, `1` and `2` points in any dimension.
+For more than `3` points, the algorithm used depends on the dimension.
+
+For the one-dimensional case we sort the points and return the minimum and maximum points,
+in that order.
+
+The two-dimensional case is handled with a planar convex hull algorithm. The
+following algorithms are available:
+
+* `"monotone_chain"`        -- compute the convex hull of points in the plane
+                               using Andrew's monotone chain method
+* `"monotone_chain_sorted"` -- the same as `"monotone_chain"` but assuming that
+                               the points are already sorted in counter-clockwise
+                               fashion
+
+See the reference docstring of each of those algorithms for details.
+
+The higher dimensional case is treated using the concrete polyhedra library
+`Polyhedra`, that gives access to libraries such as `CDDLib` and `ConvexHull.jl`.
+These libraries can be chosen by the setting the `backend` argument.
+
+### Notes
+
+For the in-place version use `convex_hull!` instead of `convex_hull`.
 
 ### Examples
 
@@ -40,57 +76,80 @@ julia> plot([Tuple(pi) for pi in points], seriestype=:scatter);
 julia> plot!(VPolygon(hull), alpha=0.2);
 ```
 """
-function convex_hull(points::Vector{VN}; algorithm::String="monotone_chain"
-                    )::Vector{VN} where {N<:Real, VN<:AbstractVector{N}}
-    return convex_hull!(copy(points), algorithm=algorithm)
+function convex_hull(points::Vector{VN};
+                     algorithm=default_convex_hull_algorithm(points),
+                     backend=nothing
+                     )::Vector{VN} where {N<:Real, VN<:AbstractVector{N}}
+    return convex_hull!(copy(points), algorithm=algorithm, backend=backend)
 end
 
-"""
-    convex_hull!(points::Vector{VN}; [algorithm]::String="monotone_chain"
-                )::Vector{VN} where {N<:Real, VN<:AbstractVector{N}}
-
-Compute the convex hull of points in the plane, in-place.
-
-### Input
-
-- `points`    -- list of 2D vectors (is modified)
-- `algorithm` -- (optional, default: `"monotone_chain"`) the convex hull
-                 algorithm; valid options are:
-
-    * `"monotone_chain"`
-    * `"monotone_chain_sorted"`
-
-### Output
-
-The convex hull as a list of 2D vectors with the coordinates of the points.
-
-### Notes
-
-See the non-modifying version `convex_hull` for more details.
-"""
 function convex_hull!(points::Vector{VN};
-                      algorithm::String="monotone_chain"
-                     )::Vector{VN} where {N<:Real, VN<:AbstractVector{N}}
+                      algorithm=default_convex_hull_algorithm(points),
+                      backend=nothing)::Vector{VN} where {N<:Real, VN<:AbstractVector{N}}
 
-    isempty(points) && return points
+    m = length(points)
 
-    @assert length(points[1]) == 2 "this function can only be used for 2-dimensional " *
-            "point clouds; try converting the points to a `VPolytope` first"
+    # =====================
+    # Treat special cases
+    # =====================
 
-    if length(points) == 1
+    # zero or one point
+    if m == 0 || m == 1
         return points
-    elseif length(points) == 2 # See #876
-        p1, p2 = points[1], points[2]
-        if p1 == p2  # check for redundancy
-            pop!(points)
-        elseif p1 <= p2
-            nothing
-        else
-            points[1], points[2] = p2, p1
+    end
+
+    n = length(first(points))
+
+    # two points
+    if m == 2
+        if n == 1 || n == 2
+            # special case, see #876
+            p1, p2 = points[1], points[2]
+            if p1 == p2  # check for redundancy
+                pop!(points)
+            elseif p1 <= p2
+                nothing
+            else
+                points[1], points[2] = p2, p1
+            end
         end
         return points
     end
 
+    # ===========================
+    # Dispatch for general cases
+    # ===========================
+
+    # _convex_hull_x can assume that there are at least three points
+    if n == 1
+        return _convex_hull_1d!(points)
+    elseif n == 2 
+        return _convex_hull_2d!(points, algorithm=algorithm)
+    else
+        return _convex_hull_nd!(points, backend=backend)
+    end
+end
+
+function _convex_hull_1d!(points::Vector{VN})::Vector{VN} where {N<:Real, VN<:AbstractVector{N}}
+    n = length(points)
+    sort!(points)
+    deleteat!(points, 2:n-1)
+end
+
+function _convex_hull_nd!(points::Vector{VN};
+                          backend=nothing
+                          )::Vector{VN} where {N<:Real, VN<:AbstractVector{N}}
+    V = VPolytope(points)
+    if backend == nothing
+        backend = default_polyhedra_backend(V, N)
+    end
+    Vch = remove_redundant_vertices(V, backend=backend)
+    return vertices_list(Vch)
+end
+
+function _convex_hull_2d!(points::Vector{VN};
+                          algorithm::String="monotone_chain"
+                          )::Vector{VN} where {N<:Real, VN<:AbstractVector{N}}
     if algorithm == "monotone_chain"
         return monotone_chain!(points)
     elseif algorithm == "monotone_chain_sorted"
@@ -98,6 +157,17 @@ function convex_hull!(points::Vector{VN};
     else
         error("the convex hull algorithm $algorithm is unknown")
     end
+end
+
+function convex_hull(P1::Vector{N},
+                     P2::HPoly{N};
+                     backend=default_polyhedra_backend(P1, N)) where {N}
+    @assert isdefined(@__MODULE__, :Polyhedra) "this function "*
+        "needs the package 'Polyhedra'"
+    Pch = convexhull(polyhedron(P1; backend=backend),
+                     polyhedron(P2; backend=backend))
+    removehredundancy!(Pch)
+    return convert(typeof(P1), Pch)
 end
 
 """

--- a/src/concrete_convex_hull.jl
+++ b/src/concrete_convex_hull.jl
@@ -45,7 +45,7 @@ following algorithms are available:
 See the reference docstring of each of those algorithms for details.
 
 The higher dimensional case is treated using the concrete polyhedra library
-`Polyhedra`, which gives access to libraries such as `CDDLib` and `ConvexHull.jl`.
+`Polyhedra`, that gives access to libraries such as `CDDLib` and `ConvexHull.jl`.
 These libraries can be chosen from the `backend` argument.
 
 ### Notes
@@ -113,7 +113,9 @@ function convex_hull!(points::Vector{VN};
             p1, p2 = points[1], points[2]
             if p1 == p2  # check for redundancy
                 pop!(points)
-            elseif p1 > p2
+            elseif p1 <= p2
+                nothing
+            else
                 points[1], points[2] = p2, p1
             end
         end

--- a/src/concrete_convex_hull.jl
+++ b/src/concrete_convex_hull.jl
@@ -17,7 +17,7 @@ Compute the convex hull of the given points.
 ### Input
 
 - `points`    -- list of vectors
-- `algorithm` -- (optional, default: `"monotone_chain"`) the convex hull
+- `algorithm` -- (optional, default: depends on the dimension) the convex hull
                  algorithm, see valid options below
 - `backend`   -- (optional, default: `"nothing"`) polyhedral computation backend
                  for higher-dimensional point sets
@@ -31,8 +31,7 @@ The convex hull as a list of vectors with the coordinates of the points.
 A pre-processing step treats the cases with `0`, `1` and `2` points in any dimension.
 For more than `3` points, the algorithm used depends on the dimension.
 
-For the one-dimensional case we sort the points and return the minimum and maximum points,
-in that order.
+For the one-dimensional case we return the minimum and maximum points, in that order.
 
 The two-dimensional case is handled with a planar convex hull algorithm. The
 following algorithms are available:
@@ -47,7 +46,7 @@ See the reference docstring of each of those algorithms for details.
 
 The higher dimensional case is treated using the concrete polyhedra library
 `Polyhedra`, that gives access to libraries such as `CDDLib` and `ConvexHull.jl`.
-These libraries can be chosen by the setting the `backend` argument.
+These libraries can be chosen from the `backend` argument.
 
 ### Notes
 
@@ -106,9 +105,7 @@ function convex_hull!(points::Vector{VN};
             p1, p2 = points[1], points[2]
             if p1 == p2  # check for redundancy
                 pop!(points)
-            elseif p1[1] <= p2[1]
-                nothing
-            else
+            elseif p1[1] > p2[1]
                 points[1], points[2] = p2, p1
             end
         elseif n == 2
@@ -116,9 +113,7 @@ function convex_hull!(points::Vector{VN};
             p1, p2 = points[1], points[2]
             if p1 == p2  # check for redundancy
                 pop!(points)
-            elseif p1 <= p2
-                nothing
-            else
+            elseif p1 > p2
                 points[1], points[2] = p2, p1
             end
         end
@@ -140,9 +135,8 @@ function convex_hull!(points::Vector{VN};
 end
 
 function _convex_hull_1d!(points::Vector{VN})::Vector{VN} where {N<:Real, VN<:AbstractVector{N}}
-    n = length(points)
-    sort!(points)
-    deleteat!(points, 2:n-1)
+    points[1:2] = [minimum(points), maximum(points)]
+    return resize!(points, 2)
 end
 
 function _convex_hull_nd!(points::Vector{VN};
@@ -153,8 +147,9 @@ function _convex_hull_nd!(points::Vector{VN};
         backend = default_polyhedra_backend(V, N)
     end
     Vch = remove_redundant_vertices(V, backend=backend)
-    # NOTE: the order of the points in Vch is lost, in general, by using filter!
-    return filter!(pi -> pi ∈ vertices_list(Vch), points)
+    m = length(Vch.vertices)
+    points[1:m] = Vch.vertices
+    return resize!(points, m)
 end
 
 function _convex_hull_2d!(points::Vector{VN};

--- a/test/unit_convex_hull.jl
+++ b/test/unit_convex_hull.jl
@@ -1,0 +1,70 @@
+using LazySets.Approximations: BoxDirections
+
+for N in [Float64, Rational{Int}]
+
+
+    # ====================================
+    # Concrete convex hull for point sets
+    # ====================================
+
+    # corner cases in dimension 1
+    @test convex_hull([Vector{N}(undef, 0)]) == [Vector{N}(undef, 0)]
+    @test convex_hull([[N(0)]]) == [[N(0)]]
+    @test convex_hull([[N(2)], [N(1)]]) == [[N(1)], [N(2)]]
+    @test convex_hull([[N(2)], [N(2)]]) == [[N(2)]]
+
+    # corner cases in dimension 2
+    @test convex_hull([[N(0), N(0)]]) == [[N(0), N(0)]]
+    @test convex_hull([[N(1), N(0)], [N(0), N(1)]]) == [[N(1), N(0)], [N(0), N(1)]]
+    @test convex_hull([[N(1), N(0)], [N(0), N(1)]]) == [[N(1), N(0)], [N(0), N(1)]]
+    
+    # corner cases in higher dimension
+    @test convex_hull([[N(0), N(0), N(0)]]) == [[N(0), N(0), N(0)]]
+    # there is no sorting in higher dim
+    @test convex_hull([[N(1), N(0), N(0)], [N(0), N(1), N(0)]]) == [[N(1), N(0), N(0)], [N(0), N(1), N(0)]]
+
+    # dimension 1
+    points_1D = [[N(2)], [N(1)], [N(1/2)], [N(-5)]]
+    @test convex_hull(points_1D) == [[N(-5)], [N(2)]]
+    convex_hull!(points_1D) # check in-place version
+    @test points_1D == [[N(-5)], [N(2)]]
+
+    # dimension 2
+    points_2D = [[N(1), N(0)], [N(1), N(1)], [N(0), N(1)], [N(-1), N(-1)], [N(1/2), N(1/2)]]
+    @test convex_hull(points_2D) == [[N(-1), N(-1)], [N(1), N(0)], [N(1), N(1)], [N(0), N(1)]]
+    convex_hull!(points_2D) # check in-place version
+    @test points_2D == [[N(-1), N(-1)], [N(1), N(0)], [N(1), N(1)], [N(0), N(1)]]
+
+    # higher dimension
+    if test_suite_polyhedra && N != Float32 # TODO : check why it breaks with Float32
+        points_3D = [[N(1), N(0), N(4)], [N(1), N(1), N(5)], [N(0), N(1), N(6)],
+                     [N(-1), N(-1), N(-7)], [N(1/2), N(1/2), N(-8)], [N(0), N(0), N(0)],
+                     [N(1), N(2), N(3)]]
+        @test convex_hull(points_3D) == [[N(1), N(0), N(4)], [N(1), N(1), N(5)],
+                                         [N(0), N(1), N(6)], [N(-1), N(-1), N(-7)],
+                                         [N(1/2), N(1/2), N(-8)], [N(1), N(2), N(3)]]
+        convex_hull!(points_3D) # check in-place version
+        @test points_3D == [[N(1), N(0), N(4)], [N(1), N(1), N(5)],
+                            [N(0), N(1), N(6)], [N(-1), N(-1), N(-7)],
+                            [N(1/2), N(1/2), N(-8)], [N(1), N(2), N(3)]]
+    end
+
+    # ============================
+    # Binary concrete convex hull 
+    # ============================
+    V1 = VPolygon([[N(1), N(0)], [N(1), N(1)], [N(0), N(1)]])
+    V2 = VPolygon([[N(-1), N(-1)], [N(1/2), N(1/2)]])
+    ch = convex_hull(V1, V2)
+    @test vertices_list(ch) == [[N(-1), N(-1)], [N(1), N(0)], [N(1), N(1)], [N(0), N(1)]]
+
+    if test_suite_polyhedra && N != Float32 # TODO : check why it breaks with Float32
+        V1 = VPolytope([[N(1), N(0), N(4)], [N(1), N(1), N(5)], [N(0), N(1), N(6)]])
+        V2 = VPolytope([[N(-1), N(-1), N(-7)], [N(1/2), N(1/2), N(-8)], [N(0), N(0), N(0)],
+                       [N(1), N(2), N(3)]])
+        ch = convex_hull(V1, V2)
+        @test ch isa VPolytope
+        @test vertices_list(ch) == [[N(1), N(0), N(4)], [N(1), N(1), N(5)],
+                                    [N(0), N(1), N(6)], [N(-1), N(-1), N(-7)],
+                                    [N(1/2), N(1/2), N(-8)], [N(1), N(2), N(3)]]
+    end
+end

--- a/test/unit_convex_hull.jl
+++ b/test/unit_convex_hull.jl
@@ -36,7 +36,7 @@ for N in [Float64, Rational{Int}]
     @test points_2D == [[N(-1), N(-1)], [N(1), N(0)], [N(1), N(1)], [N(0), N(1)]]
 
     # higher dimension
-    if test_suite_polyhedra && N != Float32 # TODO : check why it breaks with Float32
+    if test_suite_polyhedra && N != Float32 # no backend supporting Float32
         points_3D = [[N(1), N(0), N(4)], [N(1), N(1), N(5)], [N(0), N(1), N(6)],
                      [N(-1), N(-1), N(-7)], [N(1/2), N(1/2), N(-8)], [N(0), N(0), N(0)],
                      [N(1), N(2), N(3)]]
@@ -57,7 +57,7 @@ for N in [Float64, Rational{Int}]
     ch = convex_hull(V1, V2)
     @test vertices_list(ch) == ispermutation([[N(-1), N(-1)], [N(1), N(0)], [N(1), N(1)], [N(0), N(1)]])
 
-    if test_suite_polyhedra && N != Float32 # TODO : check why it breaks with Float32
+    if test_suite_polyhedra && N != Float32 # no backend supporting Float32
         V1 = VPolytope([[N(1), N(0), N(4)], [N(1), N(1), N(5)], [N(0), N(1), N(6)]])
         V2 = VPolytope([[N(-1), N(-1), N(-7)], [N(1/2), N(1/2), N(-8)], [N(0), N(0), N(0)],
                        [N(1), N(2), N(3)]])

--- a/test/unit_convex_hull.jl
+++ b/test/unit_convex_hull.jl
@@ -16,8 +16,8 @@ for N in [Float64, Rational{Int}]
     # corner cases in dimension 2
     @test convex_hull([[N(0), N(0)]]) == [[N(0), N(0)]]
     @test convex_hull([[N(1), N(0)], [N(0), N(1)]]) == [[N(1), N(0)], [N(0), N(1)]]
-    @test convex_hull([[N(1), N(0)], [N(0), N(1)]]) == [[N(1), N(0)], [N(0), N(1)]]
-    
+    @test convex_hull([[N(1), N(0)], [N(1), N(0)]]) == [[N(1), N(0)]]
+
     # corner cases in higher dimension
     @test convex_hull([[N(0), N(0), N(0)]]) == [[N(0), N(0), N(0)]]
     # there is no sorting in higher dim
@@ -40,13 +40,13 @@ for N in [Float64, Rational{Int}]
         points_3D = [[N(1), N(0), N(4)], [N(1), N(1), N(5)], [N(0), N(1), N(6)],
                      [N(-1), N(-1), N(-7)], [N(1/2), N(1/2), N(-8)], [N(0), N(0), N(0)],
                      [N(1), N(2), N(3)]]
-        @test convex_hull(points_3D) == [[N(1), N(0), N(4)], [N(1), N(1), N(5)],
+        @test convex_hull(points_3D) == ispermutation([[N(1), N(0), N(4)], [N(1), N(1), N(5)],
                                          [N(0), N(1), N(6)], [N(-1), N(-1), N(-7)],
-                                         [N(1/2), N(1/2), N(-8)], [N(1), N(2), N(3)]]
+                                         [N(1/2), N(1/2), N(-8)], [N(1), N(2), N(3)]])
         convex_hull!(points_3D) # check in-place version
-        @test points_3D == [[N(1), N(0), N(4)], [N(1), N(1), N(5)],
+        @test points_3D == ispermutation([[N(1), N(0), N(4)], [N(1), N(1), N(5)],
                             [N(0), N(1), N(6)], [N(-1), N(-1), N(-7)],
-                            [N(1/2), N(1/2), N(-8)], [N(1), N(2), N(3)]]
+                            [N(1/2), N(1/2), N(-8)], [N(1), N(2), N(3)]])
     end
 
     # ============================
@@ -55,7 +55,7 @@ for N in [Float64, Rational{Int}]
     V1 = VPolygon([[N(1), N(0)], [N(1), N(1)], [N(0), N(1)]])
     V2 = VPolygon([[N(-1), N(-1)], [N(1/2), N(1/2)]])
     ch = convex_hull(V1, V2)
-    @test vertices_list(ch) == [[N(-1), N(-1)], [N(1), N(0)], [N(1), N(1)], [N(0), N(1)]]
+    @test vertices_list(ch) == ispermutation([[N(-1), N(-1)], [N(1), N(0)], [N(1), N(1)], [N(0), N(1)]])
 
     if test_suite_polyhedra && N != Float32 # TODO : check why it breaks with Float32
         V1 = VPolytope([[N(1), N(0), N(4)], [N(1), N(1), N(5)], [N(0), N(1), N(6)]])
@@ -63,8 +63,8 @@ for N in [Float64, Rational{Int}]
                        [N(1), N(2), N(3)]])
         ch = convex_hull(V1, V2)
         @test ch isa VPolytope
-        @test vertices_list(ch) == [[N(1), N(0), N(4)], [N(1), N(1), N(5)],
+        @test vertices_list(ch) == ispermutation([[N(1), N(0), N(4)], [N(1), N(1), N(5)],
                                     [N(0), N(1), N(6)], [N(-1), N(-1), N(-7)],
-                                    [N(1/2), N(1/2), N(-8)], [N(1), N(2), N(3)]]
+                                    [N(1/2), N(1/2), N(-8)], [N(1), N(2), N(3)]])
     end
 end

